docs/concepts/states.md

```diff
@@ -237,6 +237,39 @@ def my_succeed_or_fail_hook(task, task_run, state):
 )
 ```
 
+#### Pass custom arguments to state change hooks
+The Prefect engine will call your hooks for you upon the state change, passing in the flow, flow run, and state objects. This means that you don't have the opportunity to pass in custom arguments to your hooks at runtime. However, there are a couple ways you can provide custom arguments to your hooks when you define them:
+
+- set default values for your hooks' additional arguments:
+```python
+from prefect import flow
+
+def my_hook(flow, flow_run, state, my_arg="custom_value"):
+    assert my_arg == "custom_value"
+
+@flow(on_completion=[my_hook])
+def lazy_flow():
+    pass
+```
+
+- define your hook to accept `**kwargs` and bind your custom arguments to the hook:
+```python
+from prefect import flow
+from prefect.utilities.callables import bind_args_to_fn
+
+def my_hook(flow, flow_run, state, **kwargs):
+    assert kwargs["my_arg"] == "custom_value"
+    assert kwargs["foo"] == "bar"
+
+hook_with_custom_defaults = bind_args_to_fn(

```

Comment:
i don't think i understand this example tbh; is there a reason users can't `curry` their own hooks like

```python
from functools import partial

hook_with_custom_defaults = partial(my_hook, foo="bar")
```

?

Reply:
they can! we just access the `__name__` which `partial` doesn't carry for you (unless I'm missing something) so we get an AttributeError. i was debating just putting an example in the docs with `partial` and populating `__name__` to show how you can do it - do you think that's better than adding a util?

Reply:
Interesting interesting - what do we use `__name__` for?

Reply:
only sending it to the debug logger [here](https://github.com/PrefectHQ/prefect/blob/42812c0884af73d344d39e812c3661a91699a554/src/prefect/engine.py#L2342)

perhaps i should just change that to something like
```
for hook in hooks:
    hook_name = (
        hook.__name__
        if hasattr(hook, "__name__")
        else hook.__class__.name # or some default maybe
    )
 ```
 which gives
 ```
 In [1]: from functools import partial

In [2]: from prefect import flow

In [3]: def foo(flow, flow_run, state, **kwargs):
   ...:     print(kwargs)
   ...:

In [4]: @flow(on_completion=[partial(foo, **dict(x=1, y=2, z=3))])
   ...: def some_flow():
   ...:     pass
   ...:

In [5]: some_flow()
23:22:50.699 | INFO    | prefect.engine - Created flow run 'vengeful-antelope' for flow 'some-flow'
23:22:51.523 | INFO    | Flow run 'vengeful-antelope' - Running hook 'partial' in response to entering state 'Completed'
{'x': 1, 'y': 2, 'z': 3}
23:22:51.532 | INFO    | Flow run 'vengeful-antelope' - Hook 'partial' finished running successfully
23:22:51.533 | INFO    | Flow run 'vengeful-antelope' - Finished in state Completed()
```